qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,306 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from base64 import urlsafe_b64decode
+from string import ascii_letters, digits
+import datetime
+from random import SystemRandom
+import functools
+from traceback import format_exception
+
+from tornado.web import RequestHandler
+
+from moi import r_client
+from qiita_core.exceptions import (IncorrectPasswordError, IncorrectEmailError,
+                                   UnverifiedEmailError)
+import qiita_db as qdb
+
+
+def login_client(client_id, client_secret=None):
+    """Login client_id, optionally with client_secret
+
+    Parameters
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
```

Comment:
I don't think this is actually correct, neither if we want to support it. According to the RFC6749:
```
client_secret
         REQUIRED.  The client secret.  The client MAY omit the
         parameter if the client secret is an empty string.
```
According to this description, the only time that the parameter can be omitted is that if the `client_secret` of the client is an empty string. Here, it will validate the client correctly as long as the client_id is the same. That means, the only thing that I need to impersonate somebody else is using his client_id, I don't need this `client_secret` to gain access to the system. All our clients are going to have `client_id` and `client_secret` and `client_secret` will never be an empty string, given the way they're generated.